Fix Map localization + Array[File] Delocalization

Horneth opened the pull request:
This is branched off of the Fix File Array branch because part of the fix is already in this branch.

mcovarr commented on src/main/scala/cromwell/binding/values/WdlValue.scala:
```diff
@@ -40,4 +40,9 @@ trait WdlValue {
   def valueString: String = toWdlString
+
+  def collect[T <: WdlValue](filterFn: PartialFunction[WdlValue, T]): Seq[T] = {
+    if(filterFn.isDefinedAt(this)) Seq(filterFn(this))
```
space after `if`

mcovarr commented on src/main/scala/cromwell/engine/backend/jes/JesBackend.scala:
```diff
@@ -84,6 +83,7 @@ object JesBackend {
           case Failure(e) => wdlValue
         }
       case array: WdlArray => array map gcsPathToLocal
+      case map: WdlMap => map map { case (k, v) => gcsPathToLocal(k) -> gcsPathToLocal(v) }
```
that is a lot of maps

mcovarr commented:
Is it possible to write a test for this or the Array equivalent?

Horneth commented:
Sure this definitely should have a test.

Horneth commented:
Added a test for WdlMap, will add another one for WdlArray in the other PR

Horneth commented on src/main/scala/cromwell/engine/backend/jes/JesBackend.scala:
```diff
@@ -217,31 +217,19 @@ class JesBackend extends Backend with LazyLogging with ProductionJesAuthenticati
 
   /**
    * Creates a set of JES inputs for a backend call.
+   * Note that duplicates input files (same gcs path) will be (re-)localized every time they are referenced.
```
Just wanted to bring some attention on this, I don't know if this is expected behavior or should be addressed ?

mcovarr replied:
I'd probably ticket this as Tech Debt.  Apart from additional time consumed, redundant localizations might result in egress charges if we're crossing zones or regions.

geoffjentry replied:
I don't understand the comment. What exactly is going on there? (I'm sure the wise @mcovarr is correct, however)

Horneth replied:
It means that if you have a call with say
```
File a
File b
```
Where a and b both point to the same gcs path, there will be one `JesInput` for each, which makes me to believe that JES will localize them both, and the second one will overwrite the first one, which is fine but unnecessary.

geoffjentry replied:
Ah gotcha. Yeah, agree w/ @mcovarr - not a big deal but we should probably be smarter about it

kshakir commented on src/main/scala/cromwell/binding/values/WdlArray.scala:
```diff
@@ -30,4 +30,8 @@ case class WdlArray(wdlType: WdlArrayType, value: Seq[WdlValue]) extends WdlValu
+
+  override def collect[T <: WdlValue](filterFn: PartialFunction[WdlValue, T]): Seq[T] = {
+    value flatMap { _.collect(filterFn) }
+  }
```
The name `collect` bothers me a little, because this method doesn't return a `WdlArray`, but a `Seq` instead. The scala collections [`collect`](http://www.scala-lang.org/api/2.11.7/index.html#scala.collection.GenTraversableLike@collect[B](pf:PartialFunction[A,B]):scala.collection.GenTraversable[B]) round trips the container type after filtering. If it's not going to return the original type, maybe call it `collectSeq`, but I'm terrible at naming.

This isn't a blocker to merging, just something to think about. The same goes for the other collect-to-Seq methods too.

Horneth replied:
Agreed, I picked that name because it takes a partial function like collect but it's not really a collect. More like a `findAll` maybe ?

geoffjentry replied:
Good catch. This isn't the first time something like this has come up. I'm not dead set against being sneaky like this but we should keep them to a dull roar and have a pretty good justification for it.

kshakir commented on src/main/scala/cromwell/binding/values/WdlValue.scala:
```diff
@@ -40,4 +40,9 @@ trait WdlValue {
+
+  def collect[T <: WdlValue](filterFn: PartialFunction[WdlValue, T]): Seq[T] = {
+    if (filterFn.isDefinedAt(this)) Seq(filterFn(this))
+    else Nil
```
This `if else` fits one line.

kshakir commented on src/test/scala/cromwell/engine/backend/jes/JesBackendSpec.scala:
```diff
@@ -1,15 +1,19 @@
 import cromwell.engine.workflow.{CallKey, WorkflowOptions}
 import cromwell.util.EncryptionSpec
 import cromwell.util.google.SimpleClientSecrets
+import org.scalatest.enablers.Containing
```
This and other unused imports should be optimized out by IntelliJ.

kshakir commented on src/main/scala/cromwell/engine/backend/jes/JesBackend.scala:
```diff
@@ -338,11 +328,18 @@ class JesBackend extends Backend with LazyLogging with ProductionJesAuthenticati
+          case (k, v) => wdlValueToGcsPath(k) -> wdlValueToGcsPath(v)
+        }
+        case file: WdlFile => toGcsPath(file)
+        case other => other
```
I'd like to see at least one `other` value passed through the various functions during a test.

Horneth replied:
This is not unit tested at all AFAIK. But I'll see if I can extract it to unit test it. In any case this code path is tested via Tyburn for sure.

kshakir replied:
Whoops, my comment above is in the wrong part of the code. I saw elsewhere in `JesBackendSpec` tests for `WdlString`, `WdlFile`, but not other types. I generally like basic sanity checks for the wildcard/default matches too.

kshakir commented:
Back to @Horneth for edits. Once those are addressed :+1: for merging from me.

mcovarr commented:
:+1: delta Khalid's points

geoffjentry commented on src/main/scala/cromwell/binding/values/WdlMap.scala:
```diff
@@ -55,10 +55,24 @@ case class WdlMap(wdlType: WdlMapType, value: Map[WdlValue, WdlValue]) extends W
+  override def collect[T <: WdlValue](filterFn: PartialFunction[WdlValue, T]): Seq[T] = {
+    val collected = value flatMap {
+      case (k, v) => Seq(k.collect(filterFn), v.collect(filterFn))
+    }
```
\n

Horneth commented:
I added more tests for the WdlArray and WdlFile cases as well as a test for JesOutput file paths being converted back to gcs paths in JesBackend if that's what you meant @kshakir 
Also renamed to `collectAsSeq`

kshakir commented:
Code looks good. A bunch of lines are still not being hit via unit tests, but going to call +0.4% a win. :+1:
